File: src/workers/ack_worker.py
from src.core.bus import rabbitmq_bus
from src.core.tenancy import set_current_tenant_id
from src.services.message_history_service import MessageHistoryService
from src.models.chat import MessageStatus
from src.core.database import SessionLocal
from src.core.ws import ws_manager
from src.schemas.whatsapp import WhatsAppAckStatus
from loguru import logger
import asyncio

class AckWorker:
    """
    Worker que processa confirmações de recebimento (ACKs) de mensagens do provider.
    Replicando o Serviço 'SaaS.Omnichannel.Services.AckTracker' do .NET.
    """
    async def start(self):
        logger.info("✔️ Iniciando Ack Tracking Worker em Segundo Plano...")
        
        # Subscreve na fila de ACKs do Gateway
        await rabbitmq_bus.subscribe(
            queue_name="message_ack_queue",
            routing_key="message.ack",
            exchange_name="messages_exchange",
            callback=self.handle_message_ack
        )

    async def handle_message_ack(self, payload: dict):
        """
        Ponto de entrada para cada confirmação (READ, DELIVERED, etc).
        """
        tenant_id = payload.get("tenant_id")
        ack_data = payload.get("ack") # Pydantic WhatsAppAck
        
        if not tenant_id or not ack_data:
            return

        set_current_tenant_id(tenant_id)
        
        external_id = ack_data.get("id")
        ws_status = ack_data.get("status") # Ex: 3 = Delivered, 4 = Read
        
        # 1. Converte Status do WhatsApp para o Modelo Interno
        new_status = MessageStatus.SENT
        if ws_status == WhatsAppAckStatus.DELIVERED:
            new_status = MessageStatus.DELIVERED
        elif ws_status == WhatsAppAckStatus.READ:
            new_status = MessageStatus.READ
        elif ws_status == WhatsAppAckStatus.ERROR:
            new_status = MessageStatus.ERROR
            
        # 2. Persistência de status exclusivamente no MongoDB
        from src.models.mongo.chat import MessageDocument

        mongo_msg = await MessageDocument.find_one(MessageDocument.external_id == external_id)
        # conversation_id = contact_phone do MongoDB (sem consulta ao Postgres)
        conversation_ref = mongo_msg.contact_phone if mongo_msg else external_id

        updated = await MessageHistoryService.update_message_status(external_id, new_status)
        
        if updated:
            # 3. Notificação RPC via WebSocket
            await ws_manager.publish_event(tenant_id, {
                "method": "update_message_status",
                "params": {
                    "external_id": external_id,
                    "conversation_id": conversation_ref,
                    "status": new_status,
                    "timestamp": ack_data.get("t")
                }
            })
            logger.debug(f"🔔 ACK publicado: Msg {external_id} → {new_status}")

ack_worker = AckWorker()

File: node-version/src/workers/ackWorker.js
const rabbitmqBus = require('../config/rabbitmq');
const { Message } = require('../models/nosql/Message');
const connectionManager = require('../websockets/connectionManager');
const logger = require('../utils/logger');

class AckWorker {
  async start() {
    logger.info('✔️ Iniciando Ack Tracking Worker em Segundo Plano...');
    
    await rabbitmqBus.subscribe('message_ack_queue', 'messages_exchange', 'message.ack', this.handleMessageAck.bind(this));
  }

  async handleMessageAck(payload) {
    const { tenant_id, ack } = payload;
    if (!tenant_id || !ack) return;

    // Equivalente ao WhatsAppAckStatus: 2 = SENT, 3 = DELIVERED, 4 = READ
    const externalId = ack.id;
    const wsStatus = ack.status; 
    const timestamp = ack.t || Date.now();

    let newStatus = 'sent';
    if (wsStatus === 3) newStatus = 'delivered';
    else if (wsStatus === 4) newStatus = 'read';
    else if (wsStatus === 5) newStatus = 'error';

    try {
      const msg = await Message.findOne({ 'external_id': externalId });
      
      let conversationRef = externalId;
      if (msg) {
        msg.status = newStatus;
        await msg.save();
        conversationRef = msg.contact_phone;
      } else {
        // Fallback pra criar se o Ack chegar antes da inserção por milisegundos? Não, Mongo tem upsert ou ignorar
        logger.warn(`ACK Recebido para mensagem desconhecida: ${externalId}`);
      }

      // Notificação RPC via WebSocket
      await connectionManager.broadcastToTenant(tenant_id.toUpperCase(), {
        method: 'update_message_status',
        params: {
          external_id: externalId,
          conversation_id: conversationRef,
          status: newStatus,
          timestamp: timestamp
        }
      });

      logger.debug(`🔔 ACK publicado: Msg ${externalId} → ${newStatus}`);

    } catch (e) {
      logger.error(`❌ Erro no AckWorker: ${e.message}`);
    }
  }
}

module.exports = new AckWorker();
